File: services/frontend/src/api/invite.ts
import axios from 'axios';

const api = axios.create({
  baseURL: '/api/invite',
});

export interface InviteRequest {
  target: string;
  message: string;
}

export interface InviteStatus {
  task_id: string;
  status: string;
  progress: number;
  total: number;
  success: number;
  failed: number;
  created_at: string;
  updated_at: string;
}

export async function startInvite(data: InviteRequest): Promise<InviteStatus> {
  const { data: response } = await api.post<InviteStatus>('/start', data);
  return response;
}

export async function getInviteStatus(): Promise<InviteStatus> {
  const { data } = await api.get<InviteStatus>('/status');
  return data;
}

export async function stopInvite(taskId: string): Promise<void> {
  await api.post(`/stop/${taskId}`);
} 

File: services/api-gateway/services/invite.py
from models.invite import InviteLog
from utils.celery import celery_app
from datetime import datetime
import uuid

class InviteService:
    def __init__(self):
        self.celery = celery_app

    async def start_invite(self, target: str, message: str, delay: int = 0):
        task_id = str(uuid.uuid4())
        
        # Create invite log
        invite_log = InviteLog(
            task_id=task_id,
            target=target,
            message=message,
            status="pending",
            created_at=datetime.utcnow()
        )
        await invite_log.save()

        # Start Celery task
        self.celery.send_task(
            "tasks.invite.start_invite",
            args=[task_id, target, message],
            countdown=delay
        )

        return {
            "task_id": task_id,
            "status": "started",
            "message": "Invite task started successfully"
        }

    async def get_status(self, task_id: str):
        invite_log = await InviteLog.get_by_task_id(task_id)
        if not invite_log:
            raise Exception("Task not found")

        return {
            "task_id": task_id,
            "status": invite_log.status,
            "progress": invite_log.progress,
            "total": invite_log.total,
            "success": invite_log.success,
            "failed": invite_log.failed,
            "created_at": invite_log.created_at.isoformat(),
            "updated_at": invite_log.updated_at.isoformat()
        }

    async def stop_invite(self, task_id: str):
        invite_log = await InviteLog.get_by_task_id(task_id)
        if not invite_log:
            raise Exception("Task not found")

        if invite_log.status not in ["pending", "running"]:
            raise Exception("Task cannot be stopped")

        # Revoke Celery task
        self.celery.control.revoke(task_id, terminate=True)

        # Update status
        invite_log.status = "stopped"
        invite_log.updated_at = datetime.utcnow()
        await invite_log.save()

        return {
            "task_id": task_id,
            "status": "stopped",
            "message": "Invite task stopped successfully"
        } 